File: pascals_triangle.py
def generate(numRows):
    pascal = []
    for i in range(numRows):
        pascal.append([0] * (i + 1))
        pascal[i][0] = 1
        for j in range(1, i):
            pascal[i][j] = pascal[i-1][j-1] + pascal[i-1][j]
        pascal[i][i] = 1
    return pascal

def test_generate():
    test_cases = [
        {
            "numRows": 1,
            "expected": [[1]],
        },
        {
            "numRows": 2,
            "expected": [
                [1],
                [1, 1],
            ],
        },
        {
            "numRows": 5,
            "expected": [
                [1],
                [1, 1],
                [1, 2, 1],
                [1, 3, 3, 1],
                [1, 4, 6, 4, 1],
            ],
        },
    ]

    for test_case in test_cases:
        result = generate(test_case["numRows"])
        if arrays_equal(result, test_case["expected"]):
            print(f"Test case passed for numRows = {test_case['numRows']}")
        else:
            print(f"Test case failed for numRows = {test_case['numRows']}")

def arrays_equal(arr1, arr2):
    if len(arr1) != len(arr2):
        return False
    
    for i in range(len(arr1)):
        if any(arr1[i][j] != arr2[i][j] for j in range(len(arr1[i]))):
            return False
    
    return True

test_generate()

File: PascalsTriangle.java
import java.util.ArrayList;
import java.util.List;

public class PascalsTriangle {
    public static List<List<Integer>> generate(int numRows) {
        List<List<Integer>> pascal = new ArrayList<>();
        for (int i = 0; i < numRows; i++) {
            List<Integer> row = new ArrayList<>();
            for (int j = 0; j <= i; j++) {
                if (j == 0 || j == i) {
                    row.add(1);
                } else {
                    row.add(pascal.get(i - 1).get(j - 1) + pascal.get(i - 1).get(j));
                }
            }
            pascal.add(row);
        }
        return pascal;
    }

    public static void main(String[] args) {
        testGenerate();
    }

    public static void testGenerate() {
        List<TestCase> testCases = new ArrayList<>();
        testCases.add(new TestCase(1, new ArrayList<List<Integer>>() {{
            add(new ArrayList<Integer>() {{
                add(1);
            }});
        }}));
        testCases.add(new TestCase(2, new ArrayList<List<Integer>>() {{
            add(new ArrayList<Integer>() {{
                add(1);
            }});
            add(new ArrayList<Integer>() {{
                add(1);
                add(1);
            }});
        }}));
        testCases.add(new TestCase(5, new ArrayList<List<Integer>>() {{
            add(new ArrayList<Integer>() {{
                add(1);
            }});
            add(new ArrayList<Integer>() {{
                add(1);
                add(1);
            }});
            add(new ArrayList<Integer>() {{
                add(1);
                add(2);
                add(1);
            }});
            add(new ArrayList<Integer>() {{
                add(1);
                add(3);
                add(3);
                add(1);
            }});
            add(new ArrayList<Integer>() {{
                add(1);
                add(4);
                add(6);
                add(4);
                add(1);
            }});
        }}));

        for (int i = 0; i < testCases.size(); i++) {
            TestCase testCase = testCases.get(i);
            List<List<Integer>> result = generate(testCase.numRows);
            if (arraysEqual(result, testCase.expected)) {
                System.out.println("Test case passed for numRows = " + testCase.numRows);
            } else {
                System.err.println("Test case failed for numRows = " + testCase.numRows);
            }
        }
    }

    public static boolean arraysEqual(List<List<Integer>> arr1, List<List<Integer>> arr2) {
        if (arr1.size() != arr2.size()) {
            return false;
        }

        for (int i = 0; i < arr1.size(); i++) {
            List<Integer> row1 = arr1.get(i);
            List<Integer> row2 = arr2.get(i);
            if (!row1.equals(row2)) {
                return false;
            }
        }

        return true;
    }

    static class TestCase {
        int numRows;
        List<List<Integer>> expected;

        TestCase(int numRows, List<List<Integer>> expected) {
            this.numRows = numRows;
            this.expected = expected;
        }
    }
}
